### PR TITLE
Avoid StackoverflowError when updating entity

### LINKED ISSUE #14840: StackOverflowError with multiple entity relations when elastic is enabled
##### **Overview of the issue**
PUT requests on certain entities fail if multiple relations are defined in JDL file and elastic is enabled. Same request is successful when elastic is disabled. 

##### **Motivation for or Use Case**
Using elastic for search indices while having multiple entity relations and being able to use the rest calls for storing entity changes without stackoverflow errors.

##### **Reproduce the error**
- create jHipster project with "jhipster import-jdl test.jh" with below attached JDL file (see "JDL definitions")
- start project with "mvnw"
- open gui and login
- select entity - "area"
- edit any of the fake data entries
- click on "save"
- error is thrown (stackoverflow error in log)

When setting "searchengine false" is chosen in the JDL (rebuilding and testing as above), the entity can be saved without error.

##### **Related issues**
haven't found any.

##### **Suggest a Fix**
no idea - but we found, that when we separate the calls to save entity in db and save elastic via two separated transactions, it seems to work. However, this then breaks the transaction concept if the elastic save fails.

##### **JHipster Version(s)**
current -> 7.0.1 (but 6.10.5 had the same issue) 

##### **JHipster configuration**

jhipster info:

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "authenticationType": "jwt",
    "cacheProvider": "hazelcast",
    "clientFramework": "angularX",
    "serverPort": "8081",
    "serviceDiscoveryType": false,
    "skipUserManagement": false,
    "withAdminUi": true,
    "baseName": "location",
    "buildTool": "maven",
    "databaseType": "sql",
    "devDatabaseType": "h2Memory",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": ["en", "de", "fr"],
    "messageBroker": false,
    "prodDatabaseType": "mariadb",
    "searchEngine": "elasticsearch",
    "skipClient": false,
    "testFrameworks": [],
    "websocket": false,
    "applicationType": "monolith",
    "packageName": "com.test",
    "packageFolder": "com/test",
    "nativeLanguage": "en",
    "jhipsterVersion": "7.0.1",
    "skipServer": false,
    "clientPackageManager": "npm",
    "dtoSuffix": "DTO",
    "entitySuffix": "",
    "reactive": false,
    "clientTheme": "none",
    "clientThemeVariant": "",
    "entities": ["Customer", "Area", "AreaType"],
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "blueprints": [],
    "otherModules": [],
    "pages": [],
    "creationTimestamp": 1619540342657,
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",
    "lastLiquibaseTimestamp": 1619540522000
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Customer {
  name String required unique
}
entity Area {
  name String required
  description String
}
entity AreaType {
  name String required unique
}
relationship OneToMany {
  AreaType{customerDefaultAreaType(name)} to Customer{defaultAreaType(name) required}
  AreaType{area(name)} to Area{areaType(name) required}
}
relationship ManyToMany {
  Customer{allAreaTypes(name)} to AreaType{customerAllAreaType(name)}
}

dto Customer, Area, AreaType with mapstruct
paginate Customer, Area, AreaType with infinite-scroll
service Customer, Area, AreaType with serviceClass
filter Customer, Area, AreaType

</pre>
</details>

##### **Environment and Tools**

openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)

git version 2.23.0.windows.1

node: v14.16.0

npm: 6.14.11

Docker version 19.03.13, build 4484c46d9d

docker-compose version 1.27.4, build 40524192

If you have a JDL please wrap it in below structure
  <details>
  <summary>JDL definitions</summary>
  <pre>
     application {
  config {
    baseName location
    serverPort 8081
    applicationType monolith
    packageName com.test
    authenticationType jwt
    prodDatabaseType mariadb
    devDatabaseType h2Memory
    buildTool maven
    searchEngine elasticsearch
    //searchEngine false
    enableHibernateCache true
    cacheProvider hazelcast
    enableTranslation true
    nativeLanguage          en
    languages      [en, de, fr]
    serviceDiscoveryType false
  }
  entities *
}

entity Customer{
    name String required unique
}

entity Area{
    name String required
    description String
}

entity AreaType{
    name String required unique
}

relationship ManyToOne{
                Area{areaType(name) required} to AreaType{area(name)}
}
relationship ManyToMany{
                Customer{allAreaTypes(name)} to AreaType{customerAllAreaType(name)}
}
relationship ManyToOne{
                Customer{defaultAreaType(name) required} to AreaType{customerDefaultAreaType(name)}
}

paginate * with infinite-scroll
filter * 
service * with serviceClass
dto * with mapstruct

  </pre>
  </details>

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
Entity that is having issues with PUT request is "Area":
<details>
<summary>Area.json file</summary>
<pre>
{
  "name": "Area",
  "fields": [
    {
      "fieldName": "name",
      "fieldType": "String",
      "fieldValidateRules": ["required"]
    },
    {
      "fieldName": "description",
      "fieldType": "String"
    }
  ],
  "relationships": [
    {
      "relationshipType": "many-to-one",
      "otherEntityName": "areaType",
      "otherEntityRelationshipName": "area",
      "relationshipValidateRules": "required",
      "relationshipName": "areaType",
      "otherEntityField": "name"
    }
  ],
  "entityTableName": "area",
  "dto": "mapstruct",
  "pagination": "infinite-scroll",
  "service": "serviceClass",
  "jpaMetamodelFiltering": true,
  "fluentMethods": true,
  "readOnly": false,
  "embedded": false,
  "applications": ["location"],
  "changelogDate": "20210427162102"
}
  </pre>
  </details>

##### **Browsers and Operating System**
Win10 with latest Patches, latest Chrome Browser

- [X] Checking this box is mandatory (this is just to show you read everything)

### LINKED ISSUE #16136: Elasticsearch Errors with circular related objects
## Example Project
https://github.com/hepaestus/test-app-one

### Reproduce the error
See The Test Project:
 https://github.com/hepaestus/test-app-one#development
Once you have the it set up. 
* Open http://localhost:9000/ and login as admin.
* Select the Cars entity in the menu.
* Click button to Edit a car.
* Attempt to save passengers to a car.

This should trigger the error and show an error in the alert area: 
```Handler dispatch failed; nested exception is java.lang.StackOverflowError```
See below for more of the error output.

##### **JHipster Version(s)**

```
test-app-one@0.0.1-SNAPSHOT /home/pete/git/test-app-one
└── generator-jhipster@7.1.0

```
##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**
INFO! Using JHipster version installed locally in current project's node_modules
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
test-app-one@0.0.1-SNAPSHOT /home/pete/git/test-app-one
└── generator-jhipster@7.1.0

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "clientFramework": "angularX",
    "serverPort": "8080",
    "serviceDiscoveryType": false,
    "skipUserManagement": false,
    "withAdminUi": true,
    "baseName": "testAppOne",
    "buildTool": "maven",
    "databaseType": "sql",
    "devDatabaseType": "mysql",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": false,
    "jhiPrefix": "jhi",
    "languages": ["en", "fr"],
    "messageBroker": false,
    "prodDatabaseType": "mysql",
    "searchEngine": "elasticsearch",
    "skipClient": false,
    "testFrameworks": [],
    "websocket": false,
    "applicationType": "monolith",
    "blueprints": [],
    "clientPackageManager": "npm",
    "clientTheme": "none",
    "creationTimestamp": 1629828268927,
    "dtoSuffix": "DTO",
    "jhipsterVersion": "7.1.0",
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",
    "nativeLanguage": "en",
    "otherModules": [],
    "packageName": "com.hepaestus.testappone",
    "packageFolder": "com/hepaestus/testappone",
    "reactive": false,
    "skipServer": false,
    "entitySuffix": "",
    "applicationIndex": 0,
    "entities": ["Car", "Person", "Shoe", "Driver"],
    "clientThemeVariant": "",
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "pages": [],
    "devServerPort": 4200,
    "lastLiquibaseTimestamp": 1629828508000
  }
}

</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Car {
  make String
  model String
  year LocalDate
}
entity Person {
  name String
}
entity Shoe {
  shoeSize Integer min(4) max(18)
}
entity Driver {
  licenseNumber String required pattern(/^[A-Z][a-z]{6}[0-9]{1}/)
}
relationship OneToOne {
  Person{user} to User
  Driver{person(name) required} to Person{driver}
}
relationship OneToMany {
  Driver{car required} to Car{driver required}
}
relationship ManyToMany {
  Car{passengers} to Person{car}
  Person{shoe} to Shoe{person}
}

dto Car, Person, Shoe, Driver with mapstruct
service Car, Person, Shoe, Driver with serviceClass
search Car, Person, Shoe, Driver with elasticsearch

</pre>
</details>

##### **Environment and Tools**

openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment (build 11.0.11+9-Ubuntu-0ubuntu2.20.10)
OpenJDK 64-Bit Server VM (build 11.0.11+9-Ubuntu-0ubuntu2.20.10, mixed mode, sharing)

git version 2.27.0

node: v16.3.0

npm: 7.20.5

Docker version 20.10.8, build 3967b7d

docker-compose version 1.29.2, build 5becea4c

No change to package.json was detected. No package manager install will be executed.
Congratulations, JHipster execution is complete!
Sponsored with ❤️  by @oktadev.

No change to package.json was detected. No package manager install will be executed.

##### **Overview of the issue**

When using elasticsearch and saving entities a error shows up. For example saving the  Car entity with a driver and passengers will throw an error:
```
Handler dispatch failed; nested exception is java.lang.StackOverflowError: Delayed StackOverflowError due to ReservedStackAccess annotated method
```
```
2021-09-02 13:28:25.979 DEBUG 256258 --- [  XNIO-1 task-3] c.h.testappone.web.rest.CarResource      : Enter: updateCar() with argument[s] = [2, CarDTO{id=2, make='one-to-one', model='Cuban Profit-focused', driver=null, passengers=[PersonDTO{id=1, name='Alaska', shoes=[]}, PersonDTO{id=2, name='utilize', shoes=[]}]}]
2021-09-02 13:28:25.980 DEBUG 256258 --- [  XNIO-1 task-3] c.h.testappone.web.rest.CarResource      : REST request to update Car : 2, CarDTO{id=2, make='one-to-one', model='Cuban Profit-focused', driver=null, passengers=[PersonDTO{id=1, name='Alaska', shoes=[]}, PersonDTO{id=2, name='utilize', shoes=[]}]}
2021-09-02 13:28:25.982 DEBUG 256258 --- [  XNIO-1 task-3] org.hibernate.SQL                        : select count(*) as col_0_0_ from car car0_ where car0_.id=?
2021-09-02 13:28:25.984 DEBUG 256258 --- [  XNIO-1 task-3] c.h.testappone.service.CarService        : Enter: save() with argument[s] = [CarDTO{id=2, make='one-to-one', model='Cuban Profit-focused', driver=null, passengers=[PersonDTO{id=1, name='Alaska', shoes=[]}, PersonDTO{id=2, name='utilize', shoes=[]}]}]
2021-09-02 13:28:25.984 DEBUG 256258 --- [  XNIO-1 task-3] c.h.testappone.service.CarService        : Request to save Car : CarDTO{id=2, make='one-to-one', model='Cuban Profit-focused', driver=null, passengers=[PersonDTO{id=1, name='Alaska', shoes=[]}, PersonDTO{id=2, name='utilize', shoes=[]}]}
2021-09-02 13:28:25.995 DEBUG 256258 --- [  XNIO-1 task-3] org.hibernate.SQL                        : select driver0_.id as id1_1_1_, driver0_.license_number as license_2_1_1_, driver0_.person_id as person_i3_1_1_, person1_.id as id1_5_0_, person1_.name as name2_5_0_ from driver driver0_ left outer join person person1_ on driver0_.person_id=person1_.id where driver0_.person_id=?
2021-09-02 13:28:25.997 DEBUG 256258 --- [  XNIO-1 task-3] org.hibernate.SQL                        : select driver0_.id as id1_1_1_, driver0_.license_number as license_2_1_1_, driver0_.person_id as person_i3_1_1_, person1_.id as id1_5_0_, person1_.name as name2_5_0_ from driver driver0_ left outer join person person1_ on driver0_.person_id=person1_.id where driver0_.person_id=?
2021-09-02 13:28:26.027 ERROR 256258 --- [  XNIO-1 task-3] c.h.testappone.service.CarService        : Exception in save() with cause = 'NULL' and exception = 'null'

java.lang.StackOverflowError: null
	at java.base/java.util.concurrent.ConcurrentMap.computeIfAbsent(ConcurrentMap.java:329)
	at org.springframework.data.util.ClassTypeInformation.from(ClassTypeInformation.java:71)
	at org.springframework.data.mapping.context.AbstractMappingContext.getPersistentEntity(AbstractMappingContext.java:201)
	at org.springframework.data.mapping.context.AbstractMappingContext.getPersistentEntity(AbstractMappingContext.java:87)
	at org.springframework.data.mapping.context.MappingContext.getRequiredPersistentEntity(MappingContext.java:73)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:636)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.getWriteComplexValue(MappingElasticsearchConverter.java:627)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperty(MappingElasticsearchConverter.java:601)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperties(MappingElasticsearchConverter.java:553)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:511)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:636)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.getWriteComplexValue(MappingElasticsearchConverter.java:627)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperty(MappingElasticsearchConverter.java:601)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperties(MappingElasticsearchConverter.java:553)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:511)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:636)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.getWriteComplexValue(MappingElasticsearchConverter.java:627)
...
at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:636)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.getWriteComplexValue(MappingElasticsearchConverter.java:627)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperty(MappingElasticsearchConverter.java:601)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperties(MappingElasticsearchConverter.java:553)

2021-09-02 13:28:26.059 ERROR 256258 --- [  XNIO-1 task-3] o.z.problem.spring.common.AdviceTraits   : Internal Server Error

org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.StackOverflowError
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1078)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:920)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:520)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:584)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:113)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:327)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:115)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:81)
...
at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperties(MappingElasticsearchConverter.java:553)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:511)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeEntity(MappingElasticsearchConverter.java:636)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.getWriteComplexValue(MappingElasticsearchConverter.java:627)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperty(MappingElasticsearchConverter.java:601)
	at org.springframework.data.elasticsearch.core.convert.MappingElasticsearchConverter.writeProperties(MappingElasticsearchConverter.java:553)

2021-09-02 13:28:26.076  WARN 256258 --- [  XNIO-1 task-3] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [org.springframework.web.util.NestedServletException: Handler dispatch failed; nested exception is java.lang.StackOverflowError]
```

##### **Motivation for or Use Case**

I would like to use elasticsearch out of the box as one would expect it to work.

##### **Reproduce the error**

https://github.com/hepaestus/test-app-one#development

See the Example Project here:
 https://github.com/hepaestus/test-app-one

##### **Related issues**

Cannot save entities due to recursion error.

##### **Suggest a Fix**

I am not sure of how to fix but I was curious of this might help:
```
@Field(type = FieldType.Nested, ignoreFields = {"children"})
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
```json
{
  "name": "Car",
  "fields": [
    {
      "fieldName": "make",
      "fieldType": "String"
    },
    {
      "fieldName": "model",
      "fieldType": "String"
    }
  ],
  "relationships": [
    {
      "relationshipType": "many-to-one",
      "otherEntityName": "driver",
      "otherEntityRelationshipName": "car",
      "relationshipName": "driver"
    },
    {
      "relationshipType": "many-to-many",
      "otherEntityName": "person",
      "otherEntityRelationshipName": "car",
      "relationshipName": "passengers",
      "ownerSide": true
    }
  ],
  "entityTableName": "car",
  "dto": "mapstruct",
  "pagination": "no",
  "service": "serviceClass",
  "jpaMetamodelFiltering": false,
  "fluentMethods": true,
  "readOnly": false,
  "embedded": false,
  "applications": ["testAppOne"],
  "searchEngine": "elasticsearch",
  "changelogDate": "20210824180528"
}
{
  "name": "Driver",
  "fields": [
    {
      "fieldName": "licenseNumber",
      "fieldType": "String"
    }
  ],
  "relationships": [
    {
      "relationshipType": "one-to-one",
      "otherEntityName": "person",
      "otherEntityRelationshipName": "driver",
      "relationshipName": "person",
      "ownerSide": true
    },
    {
      "relationshipType": "one-to-many",
      "otherEntityName": "car",
      "otherEntityRelationshipName": "driver",
      "relationshipName": "car"
    }
  ],
  "entityTableName": "driver",
  "dto": "mapstruct",
  "pagination": "no",
  "service": "serviceClass",
  "jpaMetamodelFiltering": false,
  "fluentMethods": true,
  "readOnly": false,
  "embedded": false,
  "applications": ["testAppOne"],
  "searchEngine": "elasticsearch",
  "changelogDate": "20210824180828"
}
{
  "name": "Person",
  "fields": [
    {
      "fieldName": "name",
      "fieldType": "String"
    }
  ],
  "relationships": [
    {
      "relationshipType": "one-to-one",
      "otherEntityName": "driver",
      "otherEntityRelationshipName": "person",
      "relationshipName": "driver",
      "ownerSide": false
    },
    {
      "relationshipType": "many-to-many",
      "otherEntityName": "car",
      "otherEntityRelationshipName": "passengers",
      "relationshipName": "car",
      "ownerSide": false
    },
    {
      "relationshipType": "many-to-many",
      "otherEntityName": "shoe",
      "otherEntityRelationshipName": "person",
      "relationshipName": "shoe",
      "ownerSide": false
    }
  ],
  "entityTableName": "person",
  "dto": "mapstruct",
  "pagination": "no",
  "service": "serviceClass",
  "jpaMetamodelFiltering": false,
  "fluentMethods": true,
  "readOnly": false,
  "embedded": false,
  "applications": ["testAppOne"],
  "searchEngine": "elasticsearch",
  "changelogDate": "20210824180628"
}
{
  "name": "Shoe",
  "fields": [
    {
      "fieldName": "shoeSize",
      "fieldType": "Integer"
    }
  ],
  "relationships": [
    {
      "relationshipType": "many-to-many",
      "otherEntityName": "person",
      "otherEntityRelationshipName": "shoe",
      "relationshipName": "person",
      "ownerSide": true
    }
  ],
  "entityTableName": "shoe",
  "dto": "mapstruct",
  "pagination": "no",
  "service": "serviceClass",
  "jpaMetamodelFiltering": false,
  "fluentMethods": true,
  "readOnly": false,
  "embedded": false,
  "applications": ["testAppOne"],
  "searchEngine": "elasticsearch",
  "changelogDate": "20210824180728"
}
```

##### **Browsers and Operating System**

Ubuntu and Chrome

- [X] Checking this box is mandatory (this is just to show you read everything)

### LINKED ISSUE #18729: Elasticsearch: StackoverflowError when trying to perform re-indexing based on JPA entities with relationships
##### **Overview of the issue**
The normal indexing mechanism of elastic search repos works while saving an entity into database. When trying to implement a custom reindexing mechanism (as my data is not coming in through some frontend at the moment :-/), I ran into an issue with taking entities directly from database (JPA entities) and saving them within the elastic search repository. Something like:

`searchRepo.saveAll(entityRepo.findAll())`

This lead to a StackoverflowError for entities with relationships as Elastic didn't understand to break the infinite cycle that we have due to the bidirectionality in the generated entities' relationship fields.

I tried various things and finally had some sort of success, but I still like to have this noted down somewhere to investigate further if there is a better solution or if it even makes sense to integrate something into generator code to just be able to support this kind of use case e.g. in a new blueprint).

Added 
`@Field(type = FieldType.Nested, ignoreFields = { "subField" })` on -- still produced the error

Added 
`(fetch = FetchType.LAZY)`  -- still produced the error

Added
`@Transient` worked, but cleared the relationships also in database, as I have put all my re-indexing code into a `@Transactional` annotated class. (Seems to  be cause by using the wrong `@Transient` :-| - the Spring data one is the correct one, the javax one will cause the mentioned issues)

Fetch the data from database -> map to dto  -> map back to entity -> save to search repository -- works, but looks ugly.

---

I also just found this old piece of code
https://github.com/geraldhumphries/generator-jhipster-elasticsearch-reindexer/blob/e300378cfc19c9a747b9aa2a43836f3807e369d2/generators/app/templates/src/main/java/package/service/_ElasticsearchIndexService.java#L186-L202
--> The relationships are loaded manually. Will check if that is somehow also an option...

##### **Motivation for or Use Case**

Implementation of re-indexing mechanisms would be nice to be able to do without big hassle.

##### **Reproduce the error**
TBD

##### **Related issues**
#16136 --> #18239 
According to some comment in #16136 this issue can be avoided when directly initializing the index properly with all relationships known... will check that out. --> Doesn't work. I have put the Nested type + ignoreFields everywhere and started with a fresh index, still failing.

TBD

##### **Suggest a Fix**
TBD

##### **JHipster Version(s)**
7.8.1

##### **JHipster configuration**
TBD

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**
TBD

##### **Browsers and Operating System**

- [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by OmarHawk**
when elasticsearch is enabled. We do add Spring Data
Transient annotation on the non-owning side
of relationships to break the artificial cycle introduced with the
bidirectional relationship model

Fix #18729
Fix #14840
Fix #16136

---

Please make sure the below checklist is followed for Pull Requests.

- [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [x] Tests are added where necessary
- [x] The JDL part is updated if necessary
- [x] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by OmarHawk** [comment]
> You should consider adding unit tests for this feature or extend existing with assertions about generated annotations ( under directory test/* )

True, ideally, there would be tests that check the actual functionality and not just the presence of some annotations. I'll see, what I can do easily.

File: generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.elastic_search.ejs [hunk 1]
```diff
@@ -31,3 +31,12 @@
     <%_ } _%>
   <&_ } -&>
 <%_ } -%>
+
+<%_ for (const relationship of relationships) { -%>
```
[the remark this answers is not in the document]
Comment (OmarHawk) [reply generators/entity-server/templates/src/main/java/package/domain/Entity.java.jhi.elastic_search.ejs:35]: Corrected via force-push

**Comment by OmarHawk** [comment]
Meh, due to the changes of #17488 entities with custom ids will fail compilation 👎 

Will fix this at well.

**Comment by OmarHawk** [comment]
> I mean unit tests for generator itself. And maybe also for CI

Well, during the CI here, plenty of applications are produced and their automated tests are being run, so adding this case into the application's datamodel to reproduce the case should be sufficient.

**Review by DanielFran — APPROVED** [review]

**Comment by DanielFran** [comment]
Thanks @OmarHawk for the contribution.
Do not forget to claim the bug bounty in https://github.com/jhipster/generator-jhipster/issues/14840

**Comment by OmarHawk** [comment]
> Thanks @OmarHawk for the contribution. Do not forget to claim the bug bounty in #14840

Thanks. :)
Claiming via https://opencollective.com/generator-jhipster/expenses/82218

**Comment by DanielFran** [comment]
@OmarHawk approved

**Comment by mheidt** [comment]
Are you guys certain, that @Transient is the solution?
Yes, it prevents the stack overflow. But the index is empty as well.
This is the issue:
You have a relationship chain from A->B->A
I want either A->B in my elastic index as well as B->A
A solution with transient breaks either one of them.
And a Mapstruct solution doesn't work as the generated setters are rewiring it again.

**Comment by OmarHawk** [comment]
Well, at least it avoids that the functionality is not available at all. And what we currently do is generating the transient annotation only in one direction of the relationship if I remember my thoughts about that again, so that the index is not really empty.

What this Pull Request also does not do is checking whether someone actually created a cycle in it's data model... this would also requiring to break the cycle somehow...

**Comment by mheidt** [comment]
Yes, and in most cases this solves the case, because you don't need both directions.
I am rather concerned about the rewiring of the parent relation. 
At least there should be an additional setter that is not doing it, so that I can try to make mapstruct use that one.

**Pull request merged**